landoapi/uplift.py

```diff
@@ -0,0 +1,184 @@
+
+from flask import current_app
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_release_managers(phab: PhabricatorClient) -> dict:
+    """Load the release-managers group details from Phabricator"""
+    groups = phab.call_conduit(
+        "project.search", constraints={"slugs": ["release-managers"]}
+    )
+    group = phab.single(groups, "data")
+    logger.info(f"Will request review from {group['fields']['name']} - {group['phid']}")
+    return group
+
+
+def check_approval_state(
+    phab: PhabricatorClient, revision_id: int, target_repository_name: str
+) -> dict:
+    """
+    Helper to load the Phabricator revision and its presence
+    in the source and target repository
+    * if the revision's target repository is the same as its current
+      repository, it's an approval
+    * otherwise it's an uplift request
+    """
+
+    # Load target repo from Phabricator
+    target_repo = phab.call_conduit(
+        "diffusion.repository.search",
+        constraints={"shortNames": [target_repository_name]},
+    )
+    target_repo = phab.single(target_repo, "data")
+
+    # Load base revision details from Phabricator
+    revision = phab.call_conduit(
+        "differential.revision.search", constraints={"ids": [revision_id]}
+    )
+    revision = phab.single(revision, "data")
+
+    # Lookup if this is an uplift or an approval request
+    is_approval = revision["fields"]["repositoryPHID"] == target_repo["phid"]
+    logger.info(
+        "Revision {} needs an {} request towards repository {}".format(
+            revision_id, is_approval and "approval" or "uplift", target_repository_name
+        )
+    )
+
+    return {
+        "target_repository": target_repo,
+        "revision": revision,
+        "is_approval": is_approval,
+    }
+
+
+def create_uplift_revision(
+    phab: PhabricatorClient,
+    source_revision: dict,
+    target_repository: dict,
+    form_content: str,
+):
+    """
+    Create a new revision on a repository, cloning a diff from another repo
+    """
+    # Check the target repository needs an approval
+    repos = get_repos_for_env(current_app.config.get("ENVIRONMENT"))
+    local_repo = repos.get(target_repository["fields"]["shortName"])
+    assert local_repo is not None, f"Unknown repository {target_repository}"
+    assert (
+        local_repo.approval_required is True
+    ), f"No approval required for {target_repository}"
+
+    # Load release managers group for review
+    release_managers = get_release_managers(phab)
+
+    # Find the source diff on phabricator
+    stack = request_extended_revision_data(phab, [source_revision["phid"]])
+    diff = stack.diffs[source_revision["fields"]["diffPHID"]]
+
+    # Get raw diff
+    raw_diff = phab.call_conduit("differential.getrawdiff", diffID=diff["id"])
+    assert raw_diff, "Missing raw source diff"
+
+    # Upload it on target repo
+    new_diff = phab.call_conduit(
+        "differential.createrawdiff",
```

Comment:
`differential.createrawdiff` will fail to handle all diffs. I believe changes that touch binary files won't properly be uploaded etc.

You'll need to use `differential.creatediff` which is way more of a pain :(. If you want an example of how this is done, take a look at how `moz-phab` does it in `--no-arc` mode: https://github.com/mozilla-conduit/review/blob/master/moz-phab

Reply:
I'm currently working on that. The hardest part is to build the list of Conduit changes from the raw diff.

Reply:
The PR now uses `differential.creatediff`. I had to add an [external library](https://github.com/mozilla/rust-parsepatch/) (incidentally made by @calixteman on my team) and use it to convert the raw patch into Phabricator changes (see https://github.com/mozilla-conduit/lando-api/pull/80/commits/3a6ca3e60a84e2c31bb7caaceb2c6b75bc7eb822)

It's of course heavily inspired by `moz-phab` source code, especially for the enums & payloads.